### DISCUSSION TITLE
how to define backend context type

### DISCUSSION THREAD

**Discussion by capaj** [trpc/trpc#2674] [Q&A] [answered]
There was a change in the last few versions of v10. Previously:
```ts
 export const t = initTRPC<{ // previously
   ctx: MyTRPCContext
 }>()()
```

worked ok to define the ctx type for the router.
I tried this:
```ts
export const t = initTRPC.create<{
  ctx: MyTRPCContext
}>()
```
but that just gives me 
```
Type '{ ctx: { prismaClient: PrismaClient<PrismaClientOptions, never, RejectOnNotFound | RejectPerOperation | undefined>; user?: undefined; } | { ...; }; }' has no properties in common with type 'Partial<InitOptions<{ ctx: {}; meta: {}; }>>'.ts(2559)
```

**Comment by sachinraja** [comment, marked as answer]
See https://trpc.io/docs/server/context for the new way of declaring context:
```ts
const t = initTRPC.context<MyTRPCContext>().create()
```

**Reply by capaj** [reply]
Thanks that looks much nicer than before!

Can't wait for v10 to be released.

**Comment by devidw** [comment]
How to get the modified context type after adding some middleware handlers with `.use()`?

Trying to do:

```ts
type FinalContextAfterMiddleware = inferAsyncReturnType<typeof protectedAppProcedure>
```

But this leads to:

> Type 'CreateProcedureReturnInput<{ _config: RootConfig<{ ctx: {...} }; meta: object; errorShape: any; transformer: t...' does not satisfy the constraint '(...args: any) => any'.
  Type 'ProcedureBuilder<{ _config: RootConfig<{ ctx: {...} }; meta: object; errorShape: any; transformer: typeof impo...' provides no match for the signature '(...args: any): any'

**Reply by Vap0r1ze** [reply]
If someone comes across this wanting to also do this, this is how I did it:
```ts
export function reContext<Params extends ProcedureParams>(next: () => Promise<MiddlewareResult<Params>>) {
    return function<Context extends object>() {
        type PatchedParams = Simplify<Omit<Params, '_ctx_out'> & { _ctx_out: Context }>;
        return next() as Promise<MiddlewareResult<PatchedParams>>;
    }
}
```

Usage:
```ts
export const authenticatedProcedure = publicProcedure.use(({ ctx, next }) => {
    if (!ctx.session) throw new Error('Unauthorized');
    return reContext(next)<RequireKeys<Context, 'session'>>();
});
```
